text/3516-public-private-dependencies.md

```diff
@@ -0,0 +1,425 @@
+  dependency edges. The directionality of the public dependency edges within the
+  subgraph doesn't matter. Each of these subgraphs represents an "ecosystem" of
+  crates publicly depending on each other. These subgraphs are "pools of public
+  types" where if you have access to the subgraph, you have access to all types
+  within that pool of types.
+* We can place a constraint that each of these "publicly connected subgraphs" are
+  required to have exactly one version of all crates internally. For example, each
+  subgraph can only have one version of Hyper.
+* Finally, we can consider all pairs of edges coming out of one node in the
+  resolution graph. If the two edges point to *two distinct publicly connected
+  subgraphs from above* and those subgraphs contain two different versions of the
+  same crate, we consider that an error. This basically means that if you privately
+  depend on Hyper 0.3 and Hyper 0.4, that's an error.
+
+If we want to go this route, some hurdles to overcome include:
+- Difficulties in working with cargo's resolver as this has been the main hang-up for stabilization over the last 6 years since the [RFC 1977] was approved
+  - For more on the complexity involved, see the thread starting at [this comment](https://github.com/rust-lang/rust/issues/44663#issuecomment-881965668)
+- More thought is needed as we found that making a dependency `public = true` can be a breaking change if the caller also depends on it but with a different semver incompatible version
+- More thought is needed on what happens if you have multiple versions of a package that are public (via renaming like `tokio_03` and `tokio_1`)
+
+### Caller-declared relations
+
+As an alternative, when declaring dependencies,
+a user could [explicitly delegate the version requirement to another package](https://github.com/rust-lang/cargo/issues/4641)
+
+One possible approach for this:
+```toml
+[package]
+name = "some-cli"
+
+[dependencies]
+clap = { from = ["clap_complete", "clap_mangen"] }
+clap_complete = "3.4"
+clap_mangen = "3.4"
+```
+When resolving the dependencies for `some-cli`,
+the resolver will not explicitly choose a version for `clap` but will continue resolving the graph.
+Upon completion, it will look to see what instance of `clap_complete` was
+resolved and act as if that was what was specified inside of the in-memory
+`clap` dependency.
+
+The package using `from` must be a public dependency of the `from` package.
+In this case, `clap` must be a public dependency of `clap_complete`.
+If the different packages in `from` do not agree on what the package
+version should resolve to (clap 3.4 vs clap 4.0), then it is an error.
+
+Compared to the resolver doing this implicitly
+- It is unclear if this would be any more difficult to implement in the resolver
+- Changing a dependency from `public = false` to `public = true` is backwards compatible because it has no affect on existing callers.
```

Comment:
```suggestion
- Changing a dependency from `public = false` to `public = true` is backwards compatible because it has no effect on existing callers.
```